client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -527,18 +528,31 @@ void trackFailure(Exception exception) {
 
     private static class SyncResponseListener implements ResponseListener {
         final CountDownLatch latch = new CountDownLatch(1);
-        volatile Response response;
-        volatile Exception exception;
+        AtomicReference<Response> response = new AtomicReference<>();
+        AtomicReference<Exception> exception = new AtomicReference<>();
 
         @Override
         public void onSuccess(Response response) {
-            this.response = response;
+            boolean wasResponseNull = this.response.compareAndSet(null, response);
+            if (wasResponseNull == false) {
+                throw new IllegalStateException("response is already set");
+            }
             latch.countDown();
         }
 
         @Override
         public void onFailure(Exception exception) {
-            this.exception = exception;
+            while (true) {
```

Comment:
I will change this to throw the exception rather than keeping track of it. if we track it it will get lost as the countDown has already happened and nobody would be waiting for it. If I let it bubble up at least our tests would catch it.